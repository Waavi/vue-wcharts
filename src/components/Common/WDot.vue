<template>
    <WTrans
        :initialProps="{
            r: 0
        }"
        :transition="transition"
    >
        <circle
            :cx="x"
            :cy="y"
            :stroke="styles.stroke"
            :fill="styles.fill"
            :r="rStyle"
            :stroke-width="styles.strokeWidth"
            :style="{ opacity: styles.opacity, cursor: styles.cursor }"
            @click="$emit('onClick', info)"
            @mouseenter="handleMouseEnter"
            @mouseleave="Chart.cleanActive"
        />
    </WTrans>
</template>

<script>
import VueTypes from 'vue-types'
import { WTrans } from '../../transitions'

export default {
    name: 'WDot',
    type: 'cartesian',
    inject: ['Chart'],
    components: {
        WTrans,
    },
    props: {
        index: VueTypes.number.isRequired,
        cartesianIndex: VueTypes.number.isRequired,
        x: VueTypes.number.isRequired,
        y: VueTypes.number.isRequired,
        styles: VueTypes.shape({
            fill: VueTypes.string,
            stroke: VueTypes.string,
            strokeWidth: VueTypes.number,
            radius: VueTypes.number,
            hoverRadius: VueTypes.number,
            opacity: VueTypes.number,
            cursor: VueTypes.string.def('default'),
        }).isRequired,
        transition: VueTypes.string.isRequired,
        info: VueTypes.shape({
            id: VueTypes.any,
            label: VueTypes.any,
            value: VueTypes.array,
        }).loose,
    },
    computed: {
        // Return styles active or default point,
        rStyle () {
            if (!this.Chart.active.el) return this.styles.radius
            const { el } = this.Chart.active
            return el.id === this.cartesianIndex && el.point === this.index ? this.styles.hoverRadius : this.styles.radius
        },
    },
    methods: {
        // Set active element
        handleMouseEnter (event) {
            const { setActive } = this.Chart
            setActive(
                this.info,
                event
            )
        },
    },
}
</script>
