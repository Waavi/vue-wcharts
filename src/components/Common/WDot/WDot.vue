<template>
    <WTrans
        :initialProps="{
            r: 0
        }"
        :transition="transition"
    >
        <circle
            :cx="x"
            :cy="y"
            :stroke="stroke"
            :stroke-width="strokeWidth || themeStyles.strokeWidth"
            :fill="fill"
            :r="rStyle"
            :style="dotStyles"
            v-on="dotListeners"
        />
    </WTrans>
</template>

<script>
import VueTypes from 'vue-types'
import merge from '../../../utils/merge'
import { WTrans } from '../../../transitions'
import themeMixin from '../../../mixins/theme'

export default {
    name: 'WDot',
    type: 'cartesian',
    inject: ['Chart'],
    components: {
        WTrans,
    },
    mixins: [themeMixin],
    props: {
        index: VueTypes.number.isRequired,
        cartesianIndex: VueTypes.number.isRequired,
        trigger: VueTypes.oneOf(['hover', 'click', 'manual']).def('hover'),
        x: VueTypes.number.isRequired,
        y: VueTypes.number.isRequired,
        fill: VueTypes.string,
        stroke: VueTypes.string,
        strokeWidth: VueTypes.number,
        radius: VueTypes.number,
        hoverRadius: VueTypes.number,
        info: VueTypes.shape({
            id: VueTypes.any,
            label: VueTypes.any,
            value: VueTypes.array,
        }).loose,
        styles: VueTypes.object.def({}),
        transition: VueTypes.string.isRequired,
    },
    computed: {
        dotStyles () {
            return {
                ...this.themeStyles.styles,
                ...this.styles,
                ...(this.trigger === 'click' ? { cursor: 'pointer' } : {}),
            }
        },
        // Event Listeners
        dotListeners () {
            return merge({}, this.$listeners, {
                click: (event) => {
                    if (this.trigger === 'click') this.handleActive(event)
                    this.$emit('onClick', this.info)
                },
                mouseenter: (event) => {
                    if (this.trigger === 'hover') this.handleActive(event)
                    this.$emit('onMouseenter', this.info)
                },
                mouseleave: () => {
                    if (['hover', 'click'].includes(this.trigger)) this.Chart.cleanActive()
                    this.$emit('onMouseleave')
                },
            })
        },
        // Return styles active or default point,
        rStyle () {
            if (!this.Chart.active.el) return this.radius
            const { el } = this.Chart.active
            return el.id === this.cartesianIndex && el.point === this.index ? (this.hoverRadius || this.themeStyles.hoverRadius) : (this.radius || this.themeStyles.radius)
        },
    },
    methods: {
        // Set active element
        handleActive (event) {
            const { setActive } = this.Chart
            setActive(
                this.info,
                event
            )
        },
    },
}
</script>
