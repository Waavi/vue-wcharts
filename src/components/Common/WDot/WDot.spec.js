import { shallowMount, mount } from '@vue/test-utils'
import WDot from './WDot.vue'

describe('Common/WDot', () => {
    const propsData = {
        index: 0,
        cartesianIndex: 1,
        x: 50,
        y: 50,
        transition: '',
    }

    const provide = {
        Chart: {
            active: {
                el: null,
                event: null,
            },
            activeElements: [],
            setActive: () => undefined,
            cleanActive: () => undefined,
        },
    }

    const defaultConfig = {
        propsData,
        provide,
    }

    it(`Should be render correctly`, () => {
        const wrapper = shallowMount(WDot, defaultConfig)
        expect(wrapper.html()).toMatchSnapshot()
    })

    it(`It emits the handleClick event`, () => {
        const wrapper = mount(WDot, defaultConfig)
        wrapper.trigger('click')
        expect(wrapper.emitted('onClick')).toHaveLength(1)
    })
})
