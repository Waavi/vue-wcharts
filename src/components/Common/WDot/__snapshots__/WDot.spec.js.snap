// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Common/WDot Should be render correctly 1`] = `
<wtrans-stub initialprops="[object Object]" transition="">
  <circle cx="50" cy="50" stroke="" stroke-width="2" fill="" r="0" style="cursor: default;"></circle>
</wtrans-stub>
`;
